mlir/lib/Transforms/CSE.cpp

```diff
@@ -35,8 +35,16 @@ using namespace mlir;
 namespace {
 struct SimpleOperationInfo : public llvm::DenseMapInfo<Operation *> {
   static unsigned getHashValue(const Operation *opC) {
+    auto *op = const_cast<Operation *>(opC);
+    // Use a custom hook if provided.
+    if (auto *interface = dyn_cast<DialectCSEInterface>(op->getDialect())) {
+      std::optional<unsigned> val = interface->getHashValue(op);
```

Comment:
I think a better way to structure this implementation would be to have CSE only query the dialect interface, and have a default implementation of the dialect interface that uses operation equivalence. That also means other users of the interface don't need to know the fallback behaviour.